crates/bevy_math/src/bounding/raycast3d.rs

```diff
@@ -1,71 +1,59 @@
 use super::{Aabb3d, BoundingSphere, IntersectsVolume};
-use crate::{Dir3, Ray3d, Vec3};
+use crate::{Dir3A, Ray3d, Vec3A};
 
 /// A raycast intersection test for 3D bounding volumes
 #[derive(Clone, Debug)]
 pub struct RayCast3d {
-    /// The ray for the test
-    pub ray: Ray3d,
+    /// The origin of the ray.
+    pub origin: Vec3A,
+    /// The direction of the ray.
+    pub direction: Dir3A,
     /// The maximum distance for the ray
     pub max: f32,
     /// The multiplicative inverse direction of the ray
-    direction_recip: Vec3,
+    direction_recip: Vec3A,
 }
 
 impl RayCast3d {
     /// Construct a [`RayCast3d`] from an origin, [`Dir3`], and max distance.
-    pub fn new(origin: Vec3, direction: Dir3, max: f32) -> Self {
-        Self::from_ray(Ray3d { origin, direction }, max)
+    pub fn new(origin: impl Into<Vec3A>, direction: impl Into<Dir3A>, max: f32) -> Self {
+        let direction = direction.into();
+        Self {
+            origin: origin.into(),
+            direction,
+            direction_recip: direction.recip(),
+            max,
+        }
     }
 
     /// Construct a [`RayCast3d`] from a [`Ray3d`] and max distance.
     pub fn from_ray(ray: Ray3d, max: f32) -> Self {
-        Self {
-            ray,
-            direction_recip: ray.direction.recip(),
-            max,
-        }
+        Self::new(ray.origin, ray.direction, max)
     }
 
     /// Get the cached multiplicative inverse of the direction of the ray.
-    pub fn direction_recip(&self) -> Vec3 {
+    pub fn direction_recip(&self) -> Vec3A {
         self.direction_recip
     }
 
     /// Get the distance of an intersection with an [`Aabb3d`], if any.
     pub fn aabb_intersection_at(&self, aabb: &Aabb3d) -> Option<f32> {
-        let (min_x, max_x) = if self.ray.direction.x.is_sign_positive() {
-            (aabb.min.x, aabb.max.x)
-        } else {
-            (aabb.max.x, aabb.min.x)
-        };
-        let (min_y, max_y) = if self.ray.direction.y.is_sign_positive() {
-            (aabb.min.y, aabb.max.y)
-        } else {
-            (aabb.max.y, aabb.min.y)
-        };
-        let (min_z, max_z) = if self.ray.direction.z.is_sign_positive() {
-            (aabb.min.z, aabb.max.z)
-        } else {
-            (aabb.max.z, aabb.min.z)
-        };
+        let positive = self.direction.signum().cmpgt(Vec3A::ZERO);
```

Comment:
is the signum needed here?

Reply:
Without the signum it'll break on 0 vs -0, signum always turns it into 1 or -1 ... Would be nice if glam had a function to get a BVec3A for the signs directly tho :thinking: